### DISCUSSION TITLE
Is sklearn pipeline via river submodules possible

### DISCUSSION THREAD

**Discussion by prabhant** [online-ml/river#553] [Q&A] [answered]
If i have something like a river scaler and classifier then is it possible to put them together in a sklearn pipeline? Or do i need to convert the entire river pipeline then only i can arrange it like that?
If its possible then can you give an example because i was not able to make it work by myself

**Comment by MaxHalford** [comment, marked as answer]
Have you checked out the `compat` module? It has some utilities for wrapping a River model to make it compatible with scikit-learn, and vice-versa.

**Reply by prabhant** [reply]
Hi, Yes i checked out the compat module. Let me explain it in a better way
If i have 3 river objects: a scaler, feature extender and a classifier then it works fine if i use River pipeline first and then convert to sklearn via river_to_sklearn function but if i convert each element separately then i cant rearrange them to a sklearn pipeline. 
```
model = Pipeline(steps=['Scaler', dataprep,
                        'Feature', featureprep,
                        'Classifier',classifier])
```
I get this error

```
/usr/local/lib/python3.7/dist-packages/sklearn/pipeline.py in _validate_steps(self)
    165 
    166     def _validate_steps(self):
--> 167         names, estimators = zip(*self.steps)
    168 
    169         # validate names

TypeError: zip argument #2 must support iteration
```

**Reply by prabhant** [reply]
Is there any possible way to do that or its not possible

**Reply by MaxHalford** [reply]
You code is not correct. You need to pass a list of tuples:

```py
model = Pipeline(steps=[
    ('Scaler', dataprep),
    ('Feature', featureprep),
    ('Classifier', classifier)
])
```

**Reply by prabhant** [reply]
Oh thanks, my bad.

**Reply by prabhant** [reply]
Is there also any way to change the hyperparameters of the converted model?

**Reply by MaxHalford** [reply]
You should be able to use `model._set_params` and provide a nested dict with the params you want. It depends on what you want to do exactly.

**Reply by medha-chippa** [reply]
@prabhant hello, did you use the river_to_sklearn() from the compact module to convert your river model to the one compatible with scikit learn. 
I've been wanting to convert a model in river to the one that is compatible with scikitlearn. But I've not been able to do this. It doesn't seem to work. Can you please guide me in this regard. 
Thank you

**Reply by prabhant** [reply]
@medha-chippa you can ping me on my email (you can find it in the github profile
)

**Reply by prabhant** [reply]
@MaxHalford I tried the model.set_params
This was my river to skl model
`River2SKLClassifier(river_estimator=LogisticRegression (
  optimizer=SGD (
    lr=Constant (
      learning_rate=0.01
    )
  )
  loss=Log (
    weight_pos=1.
    weight_neg=1.
  )
  l2=0.
  intercept_init=0.
  intercept_lr=Constant (
    learning_rate=0.01
  )
  clip_gradient=1e+12
  initializer=Zeros ()
))`

Now i wanted to change the hyperparameters like loss and l2
but when i check the keys available with get_params i get
`{'river_estimator': LogisticRegression (
   optimizer=SGD (
     lr=Constant (
       learning_rate=0.01
     )
   )
   loss=Log (
     weight_pos=1.
     weight_neg=1.
   )
   l2=0.
   intercept_init=0.
   intercept_lr=Constant (
     learning_rate=0.01
   )
   clip_gradient=1e+12
   initializer=Zeros ()
 )}`

and the only key here is river estimator. So I can't really change the l2 directly via model.set_params. Is this the intended behavior?
